src/reactor/fns.js

```diff
@@ -223,65 +226,79 @@ export function getOption(reactorState, option) {
+ * removeObserver(observerState, reactorState, [])
+ * removeObserver(observerState, reactorState, [], handler)
+ * removeObserver(observerState, reactorState, ['keyPath'])
+ * removeObserver(observerState, reactorState, ['keyPath'], handler)
+ * removeObserver(observerState, reactorState, getter)
+ * removeObserver(observerState, reactorState, getter, handler)
  * @param {ObserverState} observerState
+ * @param {ReactorState} reactorState
  * @param {KeyPath|Getter} getter
  * @param {Function} handler
- * @return {ObserverState}
+ * @return {Array}
  */
-export function removeObserver(observerState, getter, handler) {
+export function removeObserver(observerState, reactorState, getter, handler) {
   const entriesToRemove = observerState.get('observersMap').filter(entry => {
     // use the getterKey in the case of a keyPath is transformed to a getter in addObserver
-    let entryGetter = entry.get('getterKey')
+    const entryGetter = entry.get('getterKey')
+    const entryGetterLiteral = convertToGetterLiteral(entryGetter)
     let handlersMatch = (!handler || entry.get('handler') === handler)
     if (!handlersMatch) {
       return false
     }
     // check for a by-value equality of keypaths
-    if (isKeyPath(getter) && isKeyPath(entryGetter)) {
-      return isEqual(getter, entryGetter)
+    if (isKeyPath(getter) && isKeyPath(entryGetterLiteral)) {
+      return isEqual(getter, entryGetterLiteral)
     }
     // we are comparing two getters do it by reference
     return (getter === entryGetter)
   })
 
-  return observerState.withMutations(map => {
-    entriesToRemove.forEach(entry => removeObserverByEntry(map, entry))
+  // Update both observer and reactor state
+  observerState = observerState.withMutations(oState => {
+    reactorState = reactorState.withMutations(rState => {
+      entriesToRemove.forEach(entry => { removeObserverByEntry(oState, rState, entry) })
+    })
   })
+  return [
+    observerState,
+    reactorState,
+  ]
 }
 
 /**
  * Removes an observer entry by id from the observerState
  * @param {ObserverState} observerState
+ * @param {ReactorState} reactorState
  * @param {Immutable.Map} entry
- * @return {ObserverState}
+ * @return {Array}
  */
-export function removeObserverByEntry(observerState, entry) {
-  return observerState.withMutations(map => {
-    const id = entry.get('id')
-    const storeDeps = entry.get('storeDeps')
-
-    if (storeDeps.size === 0) {
-      map.update('any', anyObsevers => anyObsevers.remove(id))
-    } else {
-      storeDeps.forEach(storeId => {
-        map.updateIn(['stores', storeId], observers => {
-          if (observers) {
-            // check for observers being present because reactor.reset() can be called before an unwatch fn
-            return observers.remove(id)
-          }
-          return observers
+export function removeObserverByEntry(observerState, reactorState, entry) {
+  return [
+    observerState.withMutations(map => {
```

Comment:
putting this logic positionally in a return array is complex.

Put in a variable and return both variables.